detectron2/data/datasets/coco.py

```diff
@@ -280,6 +286,142 @@ def file2id(folder_path, file_path):
+                    num_keypoints = annotation["num_keypoints"]
+                else:
+                    num_keypoints = sum(kp > 0 for kp in keypoints[2::3])
+
+            # COCO requirement:
+            #   linking annotations to images
+            #   "id" field must start with 1
+            coco_annotation["id"] = len(coco_annotations) + 1
+            coco_annotation["image_id"] = image_dict["image_id"]
+            coco_annotation["bbox"] = bbox
+            coco_annotation["area"] = area
+            coco_annotation["category_id"] = annotation["category_id"]
+            coco_annotation["iscrowd"] = annotation.get("iscrowd", 0)
+
+            # Add optional fields
+            if "keypoints" in annotation:
+                coco_annotation["keypoints"] = keypoints
+                coco_annotation["num_keypoints"] = num_keypoints
+
+            if "segmentation" in annotation:
+                coco_annotation["segmentation"] = annotation["segmentation"]
+
+            coco_annotations.append(coco_annotation)
+
+    logger.info(
+        "Conversion finished, "
+        f"num images: {len(coco_images)}, num annotations: {len(coco_annotations)}"
+    )
+
+    info = {
+        "date_created": str(datetime.datetime.now()),
+        "description": "Automatically generated COCO json file for Detectron2.",
+    }
+    coco_dict = {
+        "info": info,
+        "images": coco_images,
+        "annotations": coco_annotations,
+        "categories": categories,
+        "licenses": None,
+    }
+    return coco_dict
+
+
+def convert_to_coco_json(dataset_name, output_folder="", allow_cached=True):
+    """
+    Converts dataset into COCO format and saves it to a json file.
+    dataset_name must be registered in DatastCatalog and in detectron2's standard format.
+
+    Args:
+        dataset_name:
+            reference from the config file to the catalogs
+            must be registered in DatastCatalog and in detectron2's standard format
+        output_folder: where json file will be saved and loaded from
+        allow_cached: if json file is already present then skip conversion
+    Returns:
+        cache_path: path to the COCO-format json file
+    """
+
+    # TODO: The dataset or the conversion script *may* change,
+    # a checksum would be useful for validating the cached data
+    cache_path = os.path.join(output_folder, f"{dataset_name}_coco_format.json")
+    PathManager.mkdirs(output_folder)
+    if os.path.exists(cache_path) and allow_cached:
+        logger.info(f"Reading cached annotations in COCO format from:{cache_path} ...")
+    else:
+        logger.info(f"Converting dataset annotations in '{dataset_name}' to COCO format ...)")
+        coco_dict = convert_to_coco_dict(dataset_name)
+
+        with PathManager.open(cache_path, "w") as json_file:
+            logger.info(f"Caching annotations in COCO format: {cache_path}")
+            json.dump(coco_dict, json_file)
+
+    return cache_path
+
+
```

Comment:
duplicate `import numpy` as line 436 since it's imported above now.

Reply:
done